pkg/api/middlewares.go

```diff
@@ -27,3 +29,52 @@ func getHTTPCacheGenNumberHeaderSetterMiddleware(cacheGenNumbersLoader *purger.T
 		})
 	})
 }
+
+// HTTPHeaderMiddleware adds specified HTTPHeaders to the request context
+type HTTPHeaderMiddleware struct {
+	TargetHeaders []string
+}
+
+// NewHttpHeaderMiddleware creates a new hTTPHeaderMiddleware given a string containing headers delimited by ~ (whitespaces in headers ignored)
+func NewHttpHeaderMiddleware(unparsedTargetHeaders string) HTTPHeaderMiddleware {
+	removeWhitespaceRegex, err := regexp.Compile(`\s`)
+	if err != nil {
+		util_log.Logger.Log("WhitespaceRegexCompilationStatus", "Failed")
+		return HTTPHeaderMiddleware{TargetHeaders: nil}
+	}
+	headersWithNoWhitespace := removeWhitespaceRegex.ReplaceAllString(unparsedTargetHeaders, "")
+
+	splitRegex, err2 := regexp.Compile(`~`)
+	if err2 != nil {
+		util_log.Logger.Log("SplitRegexCompilationStatus", "Failed")
+		return HTTPHeaderMiddleware{TargetHeaders: nil}
+	}
+	splitHeaders := splitRegex.Split(headersWithNoWhitespace, -1)
+	return HTTPHeaderMiddleware{TargetHeaders: splitHeaders}
+
+}
+
+// InjectRequestIdIntoHTTPRequest injects specified HTTPHeaders into the request context
+func (h HTTPHeaderMiddleware) InjectRequestIdIntoHTTPRequest(r *http.Request) context.Context {
+	// Check to make sure that Headers have not already been injected
+	testing, ok := r.Context().Value(util_log.RequestTargetsContextKey).([]string)
+	if ok && h.TargetHeaders != nil && testing[0] == h.TargetHeaders[0] {
+		return r.Context()
+	}
+	headerContents := make([]string, len(h.TargetHeaders))
```

Comment:
I think a map would be nicer here. we can send all data in one contextKey and we are not creating an array for a size that we are not sure will be final.

Something like 
`
headerContents = make(map[string]string)
for _, target := range h.TargetHeaders {
         headerContents[target] = r.Header.Get(target)
}
`

Reply:
I tried using a map before, but I kept having issues with getting back from the context. Tried it again and it seems to work fine now, not sure why I was having issues originally.